Questions and authors indexes

Crola1702 opened the pull request:
# Description

This PR adds the indexes for questions and authors (author profile and questions they've asked)

The only criteria to add an author is that they must have asked at least one question.

### Changes:
* Add "\<PLACEHOLDER>" to site titles instead of {ROS,GZ}
* Create /questions/index.html page
* Use preprocessor to create author profiles

<details>
<summary>
Screenshots:
</summary>

Questions index:

![image](https://github.com/user-attachments/assets/cbdb5c68-1c85-4b46-9a43-b3e9da2792f0)

User profile:

![image](https://github.com/user-attachments/assets/2236a5b7-9f9d-4f6e-887d-72d43a522e1a)


</details>

Crola1702 committed:
Update indexes to support only <PLACEHOLDER> site

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add questions index

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Sort questions by last created (tested with 7257 questions from gazebosim

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add author and creation date in question index layout

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Refactor question list to use renderer helper

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add author profile and question index

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add links to authors

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add placeholder in page_title helper

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

nuclearsandwich commented on Rules:
```diff
@@ -5,8 +5,30 @@ REDCARPET_OPTIONS = {
 
-compile '/index.md' do
+preprocess do
+  questions = @items.find_all('/question/*.md')
```
Do we need any compiled content (that is: rendered markdown in a layout) in order to show the index? If so, we can't use the preprocessor collection as it will be missing that.

Crola1702 replied:
No, we don't.

I'm trying to follow the same layout as https://answers.gazebosim.org/questions/, and I don't see any rendered content there

nuclearsandwich commented on content/index.md:
```diff
@@ -2,9 +2,9 @@
 ---
 
-# ROS and Gazebo Answers archive
+# <PLACEHOLDER> Answers archive
```
I think it's actually going to make the most sense to make these indexes html (with `erb` using the `erubis` filter) and we can set the title in `nanoc.yml` 
```yaml
site_title: Gazebo Answers Archive
```
and access it via `@config[:site_title]`.

The mechanics of how we swap the value here for each build will depend on which route we take for those generally but `sed`'ing out a sigil value before generating is one straightforward way. We can also get fancy and add a [custom nanoc check](https://nanoc.app/doc/testing/#defining-custom-checks) to ensure the sigil never shows up in the final content.

nuclearsandwich commented on layouts/question.html:
```diff
@@ -10,7 +10,7 @@ <h2><%= item[:title] %></h2>
         <% end %>
         <%= yield %>
-        <div class="owner"><p>Asked by <%= item[:owner_display_name] %> on <%= pretty_date item[:creation_date] %></p></div>
+        <div class="owner"><p>Asked by <a href="/users/<%= item[:owner_id] %>"><%= item[:owner_display_name] %></a> on <%= pretty_date item[:creation_date] %></p></div>
```
Depending on how many places it pops up. Adding a link helper for users might be worthwhile.

Crola1702 replied:
It only shows for 3 times. Do you think it's enough to make a this a helper?

nuclearsandwich replied:
The first time we have to change it we'll want to add one but let's see if we never do.

nuclearsandwich commented on layouts/questions-index.html:
```diff
@@ -0,0 +1,3 @@
+<%= yield %>
+<% questions = @items.find_all('/question/*.md').sort_by { |q| DateTime.parse(q[:creation_date]) }.reverse %>
+<%= render '/question-list.html', :questions => questions %>
```
Using a layout as a partial like this is not my favorite. I think this will go away if we make the actual content file html.

Crola1702 replied:
What do you mean by making the content file html?

nuclearsandwich replied:
Sending a PR to your PR.

nuclearsandwich replied:
https://github.com/ros-infrastructure/answers-archive/pull/3

Crola1702 committed:
Add title in nanoc config

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

nuclearsandwich committed:
Filter the question index with ERB instead of markdown.

Becuase there is very little content on this page, there's not much
motivation to have markdown for that content and using erb allows us to
put the questions listing directly in the content file instead of adding
a single-purpose layout file to use as a snippet.
